x/ccv/provider/types/msg.go

```diff
@@ -83,167 +71,215 @@ func NewMsgAssignConsumerKey(consumerId string, providerValidatorAddress sdk.Val
 	}
+
 	return nil
 }
 
-// ParseConsumerKeyFromJson parses the consumer key from a JSON string,
-// this replaces deserializing a protobuf any.
-func ParseConsumerKeyFromJson(jsonStr string) (pkType, key string, err error) {
-	type PubKey struct {
-		Type string `json:"@type"`
-		Key  string `json:"key"`
+// ValidateBasic implements the sdk.HasValidateBasic interface.
+func (msg *MsgChangeRewardDenoms) ValidateBasic() error {
+	emptyDenomsToAdd := len(msg.DenomsToAdd) == 0
+	emptyDenomsToRemove := len(msg.DenomsToRemove) == 0
+	// Return error if both sets are empty or nil
+	if emptyDenomsToAdd && emptyDenomsToRemove {
+		return fmt.Errorf(
+			"invalid change reward denoms proposal: both denoms to add and denoms to remove are empty")
 	}
-	var pubKey PubKey
-	err = json.Unmarshal([]byte(jsonStr), &pubKey)
-	if err != nil {
-		return "", "", err
+
+	// Return error if a denom is in both sets
+	for _, denomToAdd := range msg.DenomsToAdd {
+		for _, denomToRemove := range msg.DenomsToRemove {
+			if denomToAdd == denomToRemove {
+				return fmt.Errorf(
+					"invalid change reward denoms proposal: %s cannot be both added and removed", denomToAdd)
+			}
+		}
 	}
-	return pubKey.Type, pubKey.Key, nil
-}
 
-func NewMsgSubmitConsumerMisbehaviour(submitter sdk.AccAddress, misbehaviour *ibctmtypes.Misbehaviour) (*MsgSubmitConsumerMisbehaviour, error) {
-	return &MsgSubmitConsumerMisbehaviour{Submitter: submitter.String(), Misbehaviour: misbehaviour}, nil
-}
+	// Return error if any denom is "invalid"
+	for _, denom := range msg.DenomsToAdd {
+		if !sdk.NewCoin(denom, math.NewInt(1)).IsValid() {
+			return fmt.Errorf("invalid change reward denoms proposal: %s is not a valid denom", denom)
+		}
+	}
+	for _, denom := range msg.DenomsToRemove {
+		if !sdk.NewCoin(denom, math.NewInt(1)).IsValid() {
+			return fmt.Errorf("invalid change reward denoms proposal: %s is not a valid denom", denom)
+		}
+	}
 
-// Route implements the sdk.Msg interface.
-func (msg MsgSubmitConsumerMisbehaviour) Route() string { return RouterKey }
+	return nil
+}
 
-// Type implements the sdk.Msg interface.
-func (msg MsgSubmitConsumerMisbehaviour) Type() string {
-	return TypeMsgSubmitConsumerMisbehaviour
+func NewMsgSubmitConsumerMisbehaviour(
+	submitter sdk.AccAddress,
+	misbehaviour *ibctmtypes.Misbehaviour,
+) (*MsgSubmitConsumerMisbehaviour, error) {
+	return &MsgSubmitConsumerMisbehaviour{
+		Submitter:    submitter.String(),
+		Misbehaviour: misbehaviour,
+	}, nil
 }
 
-// Type implements the sdk.Msg interface.
+// ValidateBasic implements the sdk.HasValidateBasic interface.
 func (msg MsgSubmitConsumerMisbehaviour) ValidateBasic() error {
-	if msg.Submitter == "" {
-		return errorsmod.Wrap(sdkerrors.ErrInvalidAddress, msg.Submitter)
-	}
 	if err := ValidateConsumerId(msg.ConsumerId); err != nil {
-		return err
+		return errorsmod.Wrapf(ErrInvalidMsgSubmitConsumerMisbehaviour, "ConsumerId: %s", err.Error())
```

Comment:
ditto: see above

Reply:
the error message returned by ValidateConsumerId contains the necessary info.